src/component.jl

```diff
@@ -184,29 +195,120 @@ Base.sortperm(c::SharedComponent) = sortperm(c.data)
     end
     return h
 end
+
+#TODO: These boundchecks etc are to ensure in the case of || in @entities_in nobody is trying to load some
+#      invalid memory. Do we really need ||?
+@generated function Base.getindex(t::Tup, ::Type{T}) where {Tup <: Tuple, T<:ComponentData}
+    id = findfirst(x -> eltype(x) == T, Tup.parameters)
+    # return :(unsafe_load(@inbounds getindex(t, $id))::T) # without ||
+    return quote
+        $(Expr(:meta, :inline))
+        $(Expr(:meta, :propagate_inbounds))
+        ptr = getindex(t, $id)
+        @boundscheck if ptr === Ptr{T}()
+            throw(BoundsError(t, T))
+        end
+        unsafe_load(ptr)::T
+    end
+end
+
+@generated function Base.setindex!(t::Tup, x::T, ::Type{T}) where {Tup <: Tuple, T<:ComponentData}
+    id = findfirst(x -> eltype(x) == T, Tup.parameters)
+    # return :(unsafe_store!(@inbounds getindex(t, $id), x)) # without ||
+    return quote
+        $(Expr(:meta, :inline))
+        $(Expr(:meta, :propagate_inbounds))
+        ptr = getindex(t, $id)
+        @boundscheck if ptr === Ptr{T}()
+            throw(BoundsError(t, T))
+        end
+        unsafe_store!(ptr, x)
+    end
+end
 ########################################
 #                                      #
 #            Iteration                 #
 #                                      #
 ########################################
-struct EntityIterator{T <: Union{IndicesIterator,Indices,AbstractGroup}}
+struct EntityIterator{T <: Union{IndicesIterator,Indices,AbstractGroup}, TT <: Tuple}
     it::T
+    components::TT
 end
 
-Base.eltype(::EntityIterator) = Entity
+Base.eltype(::EntityIterator{T, TT}) where {T, TT} = EntityState{Tuple{map(x->Ptr{eltype(x)}, TT.parameters)...}}
 Base.IteratorSize(i::EntityIterator) = Base.IteratorSize(i.it)
 Base.length(i::EntityIterator) = length(i.it)
 
-function Base.iterate(i::EntityIterator, state = 1)
+struct EntityState{TT<:Tuple} <: AbstractEntity
+    e::Entity
+    pointers::TT
+end
+
+@generated function Base.getproperty(e::EntityState{TT}, f::Symbol) where {TT}
+    fn_to_DT = Dict{Symbol, DataType}()
+    ex = :(getfield(e, f))
+    ex = Expr(:elseif, :(f === :id), :(return getfield(getfield(e, :e), :id)::Int), ex)
+    for PDT in TT.parameters
+        DT = eltype(PDT)
```

Comment:
So this gets the `eltype` of the pointer? (I'm on the lookout for things which violate the `@generated` rules, but I think this is safe)

Reply:
Yes